pkgs/build-support/rust/build-rust-package/default.nix

```diff
@@ -15,61 +15,50 @@
-, buildInputs ? []
-, nativeBuildInputs ? []
-, cargoUpdateHook ? ""
-, cargoDepsHook ? ""
-, buildType ? "release"
-, meta ? {}
-, cargoLock ? null
-, cargoVendorDir ? null
-, checkType ? buildType
-, buildNoDefaultFeatures ? false
-, checkNoDefaultFeatures ? buildNoDefaultFeatures
-, buildFeatures ? [ ]
-, checkFeatures ? buildFeatures
-, depsExtraArgs ? {}
-
-# Toggles whether a custom sysroot is created when the target is a .json file.
-, __internal_dontAddSysroot ? false
-
-# Needed to `pushd`/`popd` into a subdir of a tarball if this subdir
-# contains a Cargo.toml, but isn't part of a workspace (which is e.g. the
-# case for `rustfmt`/etc from the `rust-sources).
-# Otherwise, everything from the tarball would've been built/tested.
-, buildAndTestSubdir ? null
-, ... } @ args:
-
-assert cargoVendorDir == null && cargoLock == null
-    -> !(args ? cargoSha256 && args.cargoSha256 != null) && !(args ? cargoHash && args.cargoHash != null)
-    -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock must be set";
-assert buildType == "release" || buildType == "debug";
+let rustOverlay =
+
+finalAttrs:
 
 let
+  cargoPatches = finalAttrs.cargoPatches or [];
+  buildType = finalAttrs.buildType or "release";
+  checkType = finalAttrs.checkType or buildType;
+  cargoLock = finalAttrs.cargoLock or null;
+  cargoVendorDir = finalAttrs.cargoVendorDir or null;
+  buildNoDefaultFeatures = finalAttrs.buildNoDefaultFeatures or false;
+  checkNoDefaultFeatures = finalAttrs.checkNoDefaultFeatures or buildNoDefaultFeatures;
+  buildFeatures = finalAttrs.buildFeatures or [ ];
+  checkFeatures = finalAttrs.checkFeatures or buildFeatures;
+  depsExtraArgs = finalAttrs.depsExtraArgs or { };
+
+  # Toggles whether a custom sysroot is created when the target is a .json file.
+  __internal_dontAddSysroot = finalAttrs.__internal_dontAddSysroot or false;
+
+  # Needed to `pushd`/`popd` into a subdir of a tarball if this subdir
+  # contains a Cargo.toml, but isn't part of a workspace (which is e.g. the
+  # case for `rustfmt`/etc from the `rust-sources).
+  # Otherwise, everything from the tarball would've been built/tested.
+  buildAndTestSubdir = finalAttrs.buildAndTestSubdir or null;
 
   cargoDeps =
+    assert cargoVendorDir == null && cargoLock == null
+        -> !(finalAttrs.cargoSha256 or null != null) && !(finalAttrs.cargoHash or null != null)
+        -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock must be set";
     if cargoVendorDir != null then null
     else if cargoLock != null then importCargoLock cargoLock
     else fetchCargoTarball ({
-      inherit src srcs sourceRoot unpackPhase cargoUpdateHook;
-      name = cargoDepsName;
+      src = finalAttrs.src or "";
+      srcs = finalAttrs.srcs or null;
+      sourceRoot = finalAttrs.sourceRoot or null;
+      unpackPhase = finalAttrs.unpackPhase or null;
+      cargoUpdateHook = finalAttrs.cargoUpdateHook or "";
+      # Name for the vendored dependencies tarball
```

Comment:
Yeah, you raise a good point here and in https://github.com/NixOS/nixpkgs/pull/194475#pullrequestreview-1205428148 above. I can rewrite it in the style you are suggesting here, as there are no existing review threads etc. that would be invalidated by this. WDYT?

Reply:
Your call.  The cleanup in your d16ba39 was enough that I was able to get through the diff without losing track of what was going on.

Reply:
Ok, then I will leave it is as is for now (unless someone else objects) -- I ran into difficult-to-debug "infinite recursion" errors when trying to refactor this, which I like to avoid debugging unless necessary :smile: